### ISSUE TITLE
Child processes are orphaned when detaching debugger

### ISSUE THREAD

**Issue by justinmchase** [OmniSharp/omnisharp-vscode#4325]
## Issue Description ##
I have setup a launch json and when I press F5 it uses `Process.Start` to create some long running child procesess. If I detach the debugger my application process is killed, but the child processes are orphaned. These processes hold open fixed ports and simply pressing F5 a second time crashes the app since the ports are already open.

## Steps to Reproduce ##
I can make a reproduction but its fairly complicated... But something like:

1. Make a C# console app that shell exec's an app that opens a port:
```cs
Process.Start("kubectl", "port-forward service/example 8080").WaitForExit();
```
2. Create a standard launch.json and hit `F5`
3. `Shift+F5` (detach debugger)
4. Press `F5` again

## Expected Behavior ##
The app successfully runs a second time and child processes are able to open their respective ports.

## Actual Behavior ##
Child processes remain running and ports are locked.

## Logs ##
```
[example] Unable to listen on port 8080: Listeners failed to create with the following errors: [unable to create listener: Error listen tcp4 127.0.0.1:8080: bind: address already in use unable to create listener: Error listen tcp6 [::1]:8080: bind: address already in use]
[example] error: unable to listen on any of the requested ports: [{8080 8080}]
```

### OmniSharp log ###
N/A

### C# log ###
N/A

## Environment information ##

**VSCode version**: 1.52.0
**C# Extension**: 1.23.8

<details>
  <summary>Mono Information</summary>
  OmniSharp using built-in mono
</details>
<details>
  <summary>Dotnet Information</summary>
.NET Core SDK (reflecting any global.json):
 Version:   3.1.404
 Commit:    470f6754b3

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.15
 OS Platform: Darwin
 RID:         osx.10.15-x64
 Base Path:   /usr/local/share/dotnet/sdk/3.1.404/

Host (useful for support):
  Version: 3.1.10
  Commit:  1721e39439

.NET Core SDKs installed:
  3.1.404 [/usr/local/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.10 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.10 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>
<details><summary>Visual Studio Code Extensions</summary>

|Extension|Author|Version|
|---|---|---|
|csharp|ms-dotnettools|1.23.8|
|debugger-for-chrome|msjsdiag|4.12.11|
|githistory|donjayamanne|0.6.14|
|gitlens|eamodio|11.1.0|
|insert-unicode|brunnerh|0.10.0|
|prettify-json|mohsen1|0.0.3|
|remote-containers|ms-vscode-remote|0.154.1|
|vscode-docker|ms-azuretools|1.9.0|
|vscode-nuget-package-manager|jmrog|1.1.6|;

</details>

**Comment by PeterWone** [comment]
I wrote the "Gentle termination" issue. while it certainly is closely related, I want to remind whoever deals with this that part of the problem is that forcibly terminated processes don't run their exit handlers, dispose their resources or flush their buffers. The proposition in the gentle termination issue was that you don't get these problems when you send the process a kill signal and let it terminate itself. 

It's not always that simple, so you might need a timeout falling back to forced termination, and if code were paused in the debugger you'd have to get user consent to resume execution because sometimes that can be a bad idea with persistent consequences.

I guess my proposition is "If you _can_ use a kill signal then do so."

**Comment by gregg-miskelly** [comment]
@PeterWone "Gentile termination" would be pretty tricky to do with a debugger. There are a bunch of issues that I don't know what you do about:
* What do you do if the click the button while the debugger is in break state? You could resume the target process, but then is the user going to be surprised that the database transaction they were trying to abort actually was actually committed?
* What do you do if further breakpoints are hit?
* What do you do if the app hit an unhandled exception so it can't run a gentile termination handler?

Because of that, and because there is a much more straight forward way of handling the scenario you had in the original request (kill the node processes as part of start debugging) this seems like the easier issue to try and fix by just looking for child processes and request gentile termination for them.

**Comment by PeterWone** [comment]
@gregg-miskelly yes thorny issues. 

For debugger paused code, you really need to consult the user. 

---
### Debug stop

Termination options are

* Resume code execution and send a kill signal
* Skip to the end of the current method, resume and send a kill signal
* Force terminate (may orphan resources and child processes)
* Cancel

[Resume] [Skip resume] [Force] [Cancel]

**Comment by gregg-miskelly** [comment]
Yup, but there is no way for a debugger to put up UI in VS Code.

**Comment by PeterWone** [comment]
OK, so you need some code between the click handler and the code that tells the debugger to stop the process. On the face of it this is not a complex change. (Famous last words).

**Comment by justinmchase** [comment]
You probably don't need to ask the user anything, I think you have enough information already; at least in VSCode you would. The debugger is either attaching to a running process or launching a process to be debugged.

When attached it offers the button "disconnect" and when that is clicked you would do nothing, the process was launched and is being managed elsewhere.

 When "launched" the UI offers you the button "Stop", which when pressed you would hard kill all the processes. When you hit stop you expect all of the processes to immediately stop and not keep running and do more work, the whole process tree should just abruptly end.
 
I would not try to ask the user what behavior they wanted in this case. You never want to "resume and send a kill signal" for example. If the user wants that they can just hit F5 to resume and then press Ctrl+C in the terminal to send the kill signal. You don't need to be clever here.

When I press Stop, it means stop. If the Stop button does anything other than stopping it will be highly annoying.

**Comment by PeterWone** [comment]
@justinmchase then leave your setting on the default of legacy behaviour, and not be inconvenienced in the slightest.

**Comment by gregg-miskelly** [comment]
@PeterWone since you are so passionate about supporting gentile termination, I will go ahead and reopen #4553 and move your comments there. This work item will continue to be about terminating child processes when we terminate the root process.

**Comment by PeterWone** [comment]
@gregg-miskelly That's very kind of you. 

I have cloned the code-base to look into the issues of implementing this. If you have any start-here pointers they would be welcome. For example, where to find the button handler code and some notes on how it's put together.

**Comment by gregg-miskelly** [comment]
@PeterWone I will follow up in the other issue.

**Comment by justinmchase** [comment]
I forgot I had previously created this issue and just went to bing and searched for "vscode stop debugging orphans processes" and found my own issue :)

I guess this is still an active issue.

**Comment by SomeoneIsWorking** [comment]
This doesn't seem related to OmniSharp because I run the programs from terminal tabs and when I close VSCode, they are still in the process list

kubuntu 21.10
